src/Compiler/Driver/ParseAndCheckInputs.fs

```diff
@@ -1338,11 +1428,138 @@ let CheckClosedInputSetFinish (declaredImpls: CheckedImplFile list, tcState) =
+                        tcState,
+                        input,
+                        true
+                    )
+                    |> Cancellable.runWithoutCancellation
+
+                let priorErrors = checkForErrors2 ()
+                partialResult, (tcState, priorErrors))
+
+        // Do the parallel phase, checking all implementation files that did have a signature, in parallel.
+        let unfinishedResults =
+
+            List.zip partialResults inputsWithLoggers
+            |> List.toArray
+            |> ArrayParallel.map (fun (partialResult, (_, logger)) ->
+                use _ = UseDiagnosticsLogger logger
+                use _ = UseBuildPhase BuildPhase.TypeCheck
+
+                RequireCompilationThread ctok
+
+                match partialResult with
+                | Choice1Of2 result -> Choice1Of2 result
+                | Choice2Of2 (amap, conditionalDefines, rootSig, priorErrors, file, tcStateForImplFile, ccuSigForFile) ->
+
+                    // In the first linear part of parallel checking, we use a 'checkForErrors' that checks either for errors
+                    // somewhere in the files processed prior to this one, including from the first phase, or in the processing
+                    // of this particular file.
+                    let checkForErrors2 () = priorErrors || (logger.ErrorCount > 0)
+
+                    let topAttrs, implFile, tcEnvAtEnd, createsGeneratedProvidedTypes =
+                        CheckOneImplFile(
+                            tcGlobals,
+                            amap,
+                            tcStateForImplFile.tcsCcu,
+                            tcStateForImplFile.tcsImplicitOpenDeclarations,
+                            checkForErrors2,
+                            conditionalDefines,
+                            TcResultsSink.NoSink,
+                            tcConfig.internalTestSpanStackReferring,
+                            tcStateForImplFile.tcsTcImplEnv,
+                            Some rootSig,
+                            file
+                        )
+                        |> Cancellable.runWithoutCancellation
+
+                    let result = (tcEnvAtEnd, topAttrs, Some implFile, ccuSigForFile)
+                    Choice2Of2(result, createsGeneratedProvidedTypes))
+            |> Array.toList
+
+        let createsGeneratedProvidedTypes =
+            unfinishedResults
+            |> List.exists (function
+                | Choice1Of2 _ -> false
+                | Choice2Of2 (_, flag) -> flag)
+
+        let tcState =
+            { tcState with
+                tcsCreatesGeneratedProvidedTypes = tcState.tcsCreatesGeneratedProvidedTypes || createsGeneratedProvidedTypes
+            }
+
+        let finishedResults =
+            unfinishedResults
+            |> List.map (function
+                | Choice1Of2 result -> result
+                | Choice2Of2 (result, _) -> result)
+
+        finishedResults, tcState)
```

Comment:
This could be simplified a bit by not re-wrapping the results in `Chioce`. 

```suggestion
        let results, createsGeneratedProvidedTypesFlags =

            List.zip partialResults inputsWithLoggers
            |> List.toArray
            |> ArrayParallel.map (fun (partialResult, (_, logger)) ->
                use _ = UseDiagnosticsLogger logger
                use _ = UseBuildPhase BuildPhase.TypeCheck

                RequireCompilationThread ctok

                match partialResult with
                | Choice1Of2 result -> result, false
                | Choice2Of2 (amap, conditionalDefines, rootSig, priorErrors, file, tcStateForImplFile, ccuSigForFile) ->

                    // In the first linear part of parallel checking, we use a 'checkForErrors' that checks either for errors
                    // somewhere in the files processed prior to this one, including from the first phase, or in the processing
                    // of this particular file.
                    let checkForErrors2 () = priorErrors || (logger.ErrorCount > 0)

                    let topAttrs, implFile, tcEnvAtEnd, createsGeneratedProvidedTypes =
                        CheckOneImplFile(
                            tcGlobals,
                            amap,
                            tcStateForImplFile.tcsCcu,
                            tcStateForImplFile.tcsImplicitOpenDeclarations,
                            checkForErrors2,
                            conditionalDefines,
                            TcResultsSink.NoSink,
                            tcConfig.internalTestSpanStackReferring,
                            tcStateForImplFile.tcsTcImplEnv,
                            Some rootSig,
                            file
                        )
                        |> Cancellable.runWithoutCancellation

                    let result = (tcEnvAtEnd, topAttrs, Some implFile, ccuSigForFile)
                    result, createsGeneratedProvidedTypes)
            |> Array.toList
            |> List.unzip

        let tcState =
            { tcState with
                tcsCreatesGeneratedProvidedTypes = tcState.tcsCreatesGeneratedProvidedTypes || (createsGeneratedProvidedTypesFlags |> List.exists id)
            }

        results, tcState)
```